File: backend/src/services/botConfig.ts
import type { PrismaClient } from "@prisma/client";
import { env } from "../utils/env.js";

export interface RuntimeBotConfig {
  autonomousEnabled: boolean;
  paperTrading: boolean;
  maxOrderInr: number;
  maxOpenExposureInr: number;
  minConfidence: number;
  symbols: string[];
  slFraction: number;
  tp1Fraction: number;
  skipHighRisk: boolean;
  dailyLossLimitInr: number;
  maxConsecutiveStopLosses: number;
  maxSpreadPct: number;
  minOpenInterest: number;
  updatedAt?: string;
}

const DEFAULTS: Omit<RuntimeBotConfig, "updatedAt"> = {
  autonomousEnabled: env.AUTONOMOUS_TRADING,
  paperTrading: env.PAPER_TRADING,
  maxOrderInr: env.MAX_ORDER_INR,
  maxOpenExposureInr: env.MAX_OPEN_EXPOSURE_INR,
  minConfidence: env.MIN_SIGNAL_CONFIDENCE,
  symbols: env.AUTONOMOUS_SYMBOLS.split(",")
    .map((s) => s.trim().toUpperCase())
    .filter(Boolean),
  slFraction: env.SL_FRACTION,
  tp1Fraction: env.TP1_FRACTION,
  skipHighRisk: true,
  dailyLossLimitInr: 2000,
  maxConsecutiveStopLosses: 3,
  maxSpreadPct: 0.08,
  minOpenInterest: 10,
};

type BotConfigRow = {
  autonomousEnabled: boolean;
  paperTrading: boolean;
  maxOrderInr: number;
  maxOpenExposureInr: number;
  minConfidence: number;
  symbols: string;
  slFraction: number;
  tp1Fraction: number;
  skipHighRisk: boolean;
  dailyLossLimitInr: number;
  maxConsecutiveStopLosses: number;
  maxSpreadPct: number;
  minOpenInterest: number;
  updatedAt: Date;
};

function rowToConfig(row: BotConfigRow): RuntimeBotConfig {
  return {
    autonomousEnabled: row.autonomousEnabled,
    paperTrading: row.paperTrading,
    maxOrderInr: row.maxOrderInr,
    maxOpenExposureInr: row.maxOpenExposureInr,
    minConfidence: row.minConfidence,
    symbols: row.symbols
      .split(",")
      .map((s) => s.trim().toUpperCase())
      .filter(Boolean),
    slFraction: row.slFraction,
    tp1Fraction: row.tp1Fraction,
    skipHighRisk: row.skipHighRisk,
    dailyLossLimitInr: row.dailyLossLimitInr,
    maxConsecutiveStopLosses: row.maxConsecutiveStopLosses,
    maxSpreadPct: row.maxSpreadPct,
    minOpenInterest: row.minOpenInterest,
    updatedAt: row.updatedAt.toISOString(),
  };
}

export async function ensureBotConfig(prisma: PrismaClient): Promise<RuntimeBotConfig> {
  const existing = await prisma.botConfig.findUnique({ where: { id: "default" } });
  if (existing) return rowToConfig(existing as BotConfigRow);

  const created = await prisma.botConfig.create({
    data: {
      id: "default",
      autonomousEnabled: DEFAULTS.autonomousEnabled,
      paperTrading: DEFAULTS.paperTrading,
      maxOrderInr: DEFAULTS.maxOrderInr,
      maxOpenExposureInr: DEFAULTS.maxOpenExposureInr,
      minConfidence: DEFAULTS.minConfidence,
      symbols: DEFAULTS.symbols.join(","),
      slFraction: DEFAULTS.slFraction,
      tp1Fraction: DEFAULTS.tp1Fraction,
      skipHighRisk: DEFAULTS.skipHighRisk,
      dailyLossLimitInr: DEFAULTS.dailyLossLimitInr,
      maxConsecutiveStopLosses: DEFAULTS.maxConsecutiveStopLosses,
      maxSpreadPct: DEFAULTS.maxSpreadPct,
      minOpenInterest: DEFAULTS.minOpenInterest,
    },
  });
  return rowToConfig(created as BotConfigRow);
}

export async function getBotConfig(prisma: PrismaClient): Promise<RuntimeBotConfig> {
  return ensureBotConfig(prisma);
}

export async function updateBotConfig(
  prisma: PrismaClient,
  patch: Partial<{
    autonomousEnabled: boolean;
    paperTrading: boolean;
    maxOrderInr: number;
    maxOpenExposureInr: number;
    minConfidence: number;
    symbols: string[];
    slFraction: number;
    tp1Fraction: number;
    skipHighRisk: boolean;
    dailyLossLimitInr: number;
    maxConsecutiveStopLosses: number;
    maxSpreadPct: number;
    minOpenInterest: number;
  }>,
): Promise<RuntimeBotConfig> {
  await ensureBotConfig(prisma);

  const data: Record<string, unknown> = {};
  if (patch.autonomousEnabled !== undefined) data.autonomousEnabled = patch.autonomousEnabled;
  if (patch.paperTrading !== undefined) data.paperTrading = patch.paperTrading;
  if (patch.maxOrderInr !== undefined) data.maxOrderInr = Math.max(50, Math.min(50000, patch.maxOrderInr));
  if (patch.maxOpenExposureInr !== undefined) {
    data.maxOpenExposureInr = Math.max(50, Math.min(100000, patch.maxOpenExposureInr));
  }
  if (patch.minConfidence !== undefined) {
    data.minConfidence = Math.max(0, Math.min(100, patch.minConfidence));
  }
  if (patch.symbols !== undefined) {
    const cleaned = patch.symbols.map((s) => s.toUpperCase()).filter((s) => ["BTC", "ETH", "SOL"].includes(s));
    data.symbols = (cleaned.length ? cleaned : ["BTC"]).join(",");
  }
  if (patch.slFraction !== undefined) data.slFraction = Math.max(0.05, Math.min(0.9, patch.slFraction));
  if (patch.tp1Fraction !== undefined) data.tp1Fraction = Math.max(0.05, Math.min(5, patch.tp1Fraction));
  if (patch.skipHighRisk !== undefined) data.skipHighRisk = patch.skipHighRisk;
  if (patch.dailyLossLimitInr !== undefined) {
    data.dailyLossLimitInr = Math.max(0, Math.min(500000, patch.dailyLossLimitInr));
  }
  if (patch.maxConsecutiveStopLosses !== undefined) {
    data.maxConsecutiveStopLosses = Math.max(0, Math.min(20, Math.floor(patch.maxConsecutiveStopLosses)));
  }
  if (patch.maxSpreadPct !== undefined) {
    data.maxSpreadPct = Math.max(0.01, Math.min(0.5, patch.maxSpreadPct));
  }
  if (patch.minOpenInterest !== undefined) {
    data.minOpenInterest = Math.max(0, Math.min(1_000_000, patch.minOpenInterest));
  }

  const updated = await prisma.botConfig.update({ where: { id: "default" }, data });
  return rowToConfig(updated as BotConfigRow);
}

File: mobile/lib/models/bot_config.dart
class BotConfig {
  final bool autonomousEnabled;
  final bool paperTrading;
  final double maxOrderInr;
  final double maxOpenExposureInr;
  final double minConfidence;
  final List<String> symbols;
  final double slFraction;
  final double tp1Fraction;
  final bool skipHighRisk;
  final double dailyLossLimitInr;
  final int maxConsecutiveStopLosses;
  final double maxSpreadPct;
  final double minOpenInterest;
  final bool killed;
  final bool deltaConfigured;

  BotConfig({
    required this.autonomousEnabled,
    required this.paperTrading,
    required this.maxOrderInr,
    required this.maxOpenExposureInr,
    required this.minConfidence,
    required this.symbols,
    required this.slFraction,
    required this.tp1Fraction,
    required this.skipHighRisk,
    required this.dailyLossLimitInr,
    required this.maxConsecutiveStopLosses,
    required this.maxSpreadPct,
    required this.minOpenInterest,
    required this.killed,
    required this.deltaConfigured,
  });

  factory BotConfig.fromJson(Map<String, dynamic> json) => BotConfig(
        autonomousEnabled: json['autonomousEnabled'] as bool? ?? false,
        paperTrading: json['paperTrading'] as bool? ?? true,
        maxOrderInr: (json['maxOrderInr'] as num?)?.toDouble() ?? 1000,
        maxOpenExposureInr: (json['maxOpenExposureInr'] as num?)?.toDouble() ?? 2000,
        minConfidence: (json['minConfidence'] as num?)?.toDouble() ?? 55,
        symbols: (json['symbols'] as List?)?.map((e) => e.toString()).toList() ?? const ['BTC', 'ETH'],
        slFraction: (json['slFraction'] as num?)?.toDouble() ?? 0.4,
        tp1Fraction: (json['tp1Fraction'] as num?)?.toDouble() ?? 0.5,
        skipHighRisk: json['skipHighRisk'] as bool? ?? true,
        dailyLossLimitInr: (json['dailyLossLimitInr'] as num?)?.toDouble() ?? 2000,
        maxConsecutiveStopLosses: (json['maxConsecutiveStopLosses'] as num?)?.toInt() ?? 3,
        maxSpreadPct: (json['maxSpreadPct'] as num?)?.toDouble() ?? 0.08,
        minOpenInterest: (json['minOpenInterest'] as num?)?.toDouble() ?? 10,
        killed: json['killed'] as bool? ?? false,
        deltaConfigured: json['deltaConfigured'] as bool? ?? false,
      );

  /// General PATCH — never sends paperTrading:false (use goLive).
  Map<String, dynamic> toPatch() => {
        'autonomousEnabled': autonomousEnabled,
        if (paperTrading) 'paperTrading': true,
        'maxOrderInr': maxOrderInr,
        'maxOpenExposureInr': maxOpenExposureInr,
        'minConfidence': minConfidence,
        'symbols': symbols,
        'slFraction': slFraction,
        'tp1Fraction': tp1Fraction,
        'skipHighRisk': skipHighRisk,
        'dailyLossLimitInr': dailyLossLimitInr,
        'maxConsecutiveStopLosses': maxConsecutiveStopLosses,
        'maxSpreadPct': maxSpreadPct,
        'minOpenInterest': minOpenInterest,
      };

  BotConfig copyWith({
    bool? autonomousEnabled,
    bool? paperTrading,
    double? maxOrderInr,
    double? maxOpenExposureInr,
    double? minConfidence,
    List<String>? symbols,
    double? slFraction,
    double? tp1Fraction,
    bool? skipHighRisk,
    double? dailyLossLimitInr,
    int? maxConsecutiveStopLosses,
    double? maxSpreadPct,
    double? minOpenInterest,
    bool? killed,
    bool? deltaConfigured,
  }) {
    return BotConfig(
      autonomousEnabled: autonomousEnabled ?? this.autonomousEnabled,
      paperTrading: paperTrading ?? this.paperTrading,
      maxOrderInr: maxOrderInr ?? this.maxOrderInr,
      maxOpenExposureInr: maxOpenExposureInr ?? this.maxOpenExposureInr,
      minConfidence: minConfidence ?? this.minConfidence,
      symbols: symbols ?? this.symbols,
      slFraction: slFraction ?? this.slFraction,
      tp1Fraction: tp1Fraction ?? this.tp1Fraction,
      skipHighRisk: skipHighRisk ?? this.skipHighRisk,
      dailyLossLimitInr: dailyLossLimitInr ?? this.dailyLossLimitInr,
      maxConsecutiveStopLosses: maxConsecutiveStopLosses ?? this.maxConsecutiveStopLosses,
      maxSpreadPct: maxSpreadPct ?? this.maxSpreadPct,
      minOpenInterest: minOpenInterest ?? this.minOpenInterest,
      killed: killed ?? this.killed,
      deltaConfigured: deltaConfigured ?? this.deltaConfigured,
    );
  }
}
